README.md
httpcode
=======



[![Build Status](https://travis-ci.org/sckott/httpcode.svg)](https://travis-ci.org/sckott/httpcode)

`httpcode` is a tiny R package to search for and show http code messages and description. It's a port of the Python `httpcode` library. 

`httpcode` has no dependencies.

## Installation

Stable version


```r
install.packages("httpcode")
```

Development version


```r
devtools::install_github("sckott/httpcode")
```


```r
library("httpcode")
```

## Search by http code


```r
http_code(100)
#> <Status code: 100>
#>   Message: Continue
#>   Explanation: Request received, please continue
```


```r
http_code(400)
#> <Status code: 400>
#>   Message: Bad Request
#>   Explanation: Bad request syntax or unsupported method
```


```r
http_code(503)
#> <Status code: 503>
#>   Message: Service Unavailable
#>   Explanation: The server cannot process the request due to a high load
```


```r
http_code(999)
#> Error: No description found for code: 999
```

# Fuzzy code search


```r
http_code('1xx')
#> [[1]]
#> <Status code: 100>
#>   Message: Continue
#>   Explanation: Request received, please continue
#> 
#> [[2]]
#> <Status code: 101>
#>   Message: Switching Protocols
#>   Explanation: Switching to new protocol; obey Upgrade header
#> 
#> [[3]]
#> <Status code: 102>
#>   Message: Processing
#>   Explanation: WebDAV; RFC 2518
```


```r
http_code('3xx')
#> [[1]]
#> <Status code: 300>
#>   Message: Multiple Choices
#>   Explanation: Object has several resources -- see URI list
#> 
#> [[2]]
#> <Status code: 301>
#>   Message: Moved Permanently
#>   Explanation: Object moved permanently -- see URI list
#> 
...
```


```r
http_code('30[12]')
#> [[1]]
#> <Status code: 301>
#>   Message: Moved Permanently
#>   Explanation: Object moved permanently -- see URI list
#> 
#> [[2]]
#> <Status code: 302>
#>   Message: Found
#>   Explanation: Object moved temporarily -- see URI list
```


```r
http_code('30[34]')
#> [[1]]
#> <Status code: 303>
#>   Message: See Other
#>   Explanation: Object moved -- see Method and URL list
#> 
#> [[2]]
#> <Status code: 304>
#>   Message: Not Modified
#>   Explanation: Document has not changed since given time
```

## Search by message


```r
http_search("request")
#> [[1]]
#> <Status code: 100>
#>   Message: Continue
#>   Explanation: Request received, please continue
#> 
#> [[2]]
#> <Status code: 200>
#>   Message: OK
#>   Explanation: Request fulfilled, document follows
#> 
...
```


```r
http_search("forbidden")
#> [[1]]
#> <Status code: 403>
#>   Message: Forbidden
#>   Explanation: Request forbidden -- authorization will not help
```


```r
http_search("too")
#> [[1]]
#> <Status code: 413>
#>   Message: Request Entity Too Large
#>   Explanation: Entity is too large.
#> 
#> [[2]]
#> <Status code: 414>
#>   Message: Request-URI Too Long
#>   Explanation: URI is too long.
```


```r
http_search("birds")
#> Error: No status code found for search: : birds
```


## Bugs/features?

See [issues](https://github.com/sckott/httpcode/issues)


NEWS.md
httpcode 0.1.0
===============

NEW FEATURES

* Released to CRAN.


R/cat_for_status.R
#' Use cat or dog pictures for various HTTP status codes
#'
#' @export
#' @name cat_for_status
#'
#' @param code An http status code
#' @param browse (logical) If \code{TRUE}, opens image in default browser. If
#' \code{FALSE}, returns the URL of the image.
#' @return Opens image in your default browser, or returns URL
#' @examples \donttest{
#' # give back url
#' cat_for_status(100)
#' dog_for_status(100)
#' cat_for_status(301)
#' dog_for_status(301)
#' cat_for_status(400)
#' dog_for_status(400)
#'
#' # open image in default browser
#' cat_for_status(400, browse=TRUE)
#'
#' # not found
#' # cat_for_status(555)
#' }
cat_for_status <- function(code, browse = FALSE){
  code <- as.character(code)
  if (code %in% names(status_codes)) {
    url <- sprintf('http://httpcats.herokuapp.com/%s', code)
    if (browse) browseURL(url) else url
  } else {
    stopcode("No cat code found", "")
  }
}

#' @export
#' @rdname cat_for_status
dog_for_status <- function(code, browse = FALSE) {
  code <- as.character(code)
  if (code %in% names(status_codes)) {
    url <- sprintf('http://httpstatusdogs.com/wp-content/uploads/%s.jpg', code)
    if (browse) browseURL(url) else url
  } else {
    stopcode("No dog code found", "")
  }
}


R/code.R
print_filtered_codes <- function(code){
  code2 <- paste0(gsub("x", "\\\\d", code), "$")
  found_codes <- nozero(sapply(names(status_codes), function(x) grep(code2, x, value=TRUE)))
  if(length(found_codes) == 0) stopcode('No code found corresponding to', code)
  print_codes(found_codes)
}

print_codes <- function(codes=names(status_codes)) lapply(sort(codes), print_code)

print_code <- function(code){
  twocodes <- status_codes[code]
  if(length(twocodes[[1]]) != 2) stopcode('No description found for code', code)
  structure(msg_list(code, twocodes[[1]][[1]], twocodes[[1]][[2]]), class="http_code")
}

#' @export
print.http_code <- function(x, ...){
  cat(sprintf("<Status code: %s>", x$status_code), sep="\n")
  cat(sprintf("  Message: %s", x$message), sep="\n")
  cat(sprintf("  Explanation: %s", x$explanation), sep="\n")
}

print_search <- function(text){
  found_codes <- vapply(status_codes, function(x){
    any(vapply(x, function(y) grepl(text, y, ignore.case = TRUE), logical(1)))
  }, logical(1))
  if(any(found_codes)) print_codes(names(status_codes[found_codes])) else stopcode('No status code found for search: ', text)
}

#' Find out about http status codes
#'
#' @name http
#' @param code (character) An http status code, or a regex search for HTTP
#' status codes.
#' @param text (character) A text string to search the messages or descriptions
#' of HTTP status codes.
#' @examples
#' # search by code
#' http_code(100)
#' http_code(400)
#' http_code(503)
#'
#' # fuzzy code search
#' http_code('1xx')
#' http_code('3xx')
#' http_code('30[12]')
#' http_code('30[34]')
#' http_code('30[34]')
#'
#' # search by text message
#' http_search("request")
#' http_search("forbidden")
#' http_search("too")
#'
#' @examples \dontrun{
#' http_search("birds")
#' http_code(999)
#' }

#' @export
#' @rdname http
http_code <- function(code = NULL){
  code <- as.character(code)
  if(is.null(code)){
    print_codes()
  } else {
    if(is_three_digit_code(code)) print_code(code) else print_filtered_codes(code)
  }
}

#' @export
#' @rdname http
http_search <- function(text = NULL) print_search(text)
